### ISSUE TITLE
Support plugins on a per endpoint (and HTTP method?) basis

### ISSUE THREAD

**Issue by Nijikokun** [Mashape/kong#160]
Right now plugins encompass the entire API, meaning that they are invoked for every endpoint.

An amazing feature would be to allow specifying specific endpoints that the plugin works for, should no endpoints be specified it encompasses the entire API.

This shouldn't be done on a *per plugin basis* as that is not a DRY approach and would require each plugin to do additional work. The check should happen **before** the plugin is invoked.

**Comment by sinzone** [comment]
should be in the KONG RESTful interface?

**Comment by Nijikokun** [comment]
This also means allowing to configure the **same** plugin for a **single** API *multiple times*.

Examples:

- CORS Config `A` on API `A` on endpoints `A, D`
- CORS Config `B` on API `A` on endpoints `B, C`
- Keyauth Config `A` on API `A` on endpoints `A, C`
- Keyauth Config `A` on API `A` on endpoints `D`
- Queryauth Config `A` on API `A` on endpoints `B`

**Comment by montanaflynn** [comment]
+1

**Comment by Nijikokun** [comment]
I should also add that plugins could also be per **method** as well.

**Comment by sinzone** [comment]
@thefosk is this happening in the coming 0.3?

**Comment by thefosk** [comment]
Not in `0.3.0`, but could happen in `0.3.1`or subsequent release

**Comment by sinzone** [comment]
+1

**Comment by morphomatt** [comment]
+1

**Comment by thefosk** [comment]
This issue should also enable plugins not just per endpoint, but per HTTP method too. Like:

```shell
curl 127.0.0.1:8001/apis/test.com/plugins/
     -d "name=keyauth" \
     -d "path=/hello/*/world/*" \
     -d "method=POST"
```

**Comment by montanaflynn** [comment]
What's the status on this? I see it was removed from the 0.3.1 milestone, is it still a planned feature?

One more thing to be brought up is support for wildcards or regex when applying to endpoints.

**Comment by sinzone** [comment]
@montanaflynn the milestones are here now: https://github.com/Mashape/kong/wiki

**Comment by thefosk** [comment]
Closed in favor of #505.
